### PR TITLE
Deal 1429 side panel improvement

### PR DESCRIPTION
**Pull request by pearlanre**
## PR Checklist

Storybook can be viewed [here](DOCSPOT_URL)

- Manually tested across supported browsers

  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari

- [ ] Unit tests written (`common` at minimum)
- [ ] PR has one of the `semver-` labels
- [ ] Two core team engineer approvals
- [ ] One core team UX approval

### CODE REVIEW

File: src/components/SidePanel/SidePanel.tsx [hunk 1]
```diff
@@ -134,6 +141,12 @@ class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
 			Sets the initial width in pixels. The actual width may change if the user
 			resizes it.
 		`,
+		minWidth: number`
+		Sets the minimum width of the Side Panel.
+		`,
+		maxWidth: number`
```
Comment (jondlm) [inline src/components/SidePanel/SidePanel.tsx:147]: As I'm looking at this I'm realizing what would be _really_ nice if we could prevent the panel from going off the left side of the screen. Would it be possible to support something like `100%`? I haven't dug into this component so I don't even know if it's technically feasible.

File: src/components/SidePanel/SidePanel.tsx [hunk 2]
```diff
@@ -156,6 +169,8 @@ class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
 		preventBodyScroll: false,
 		topOffset: 0,
 		width: 240,
+		minWidth: null,
```
Comment (jondlm) [inline src/components/SidePanel/SidePanel.tsx:172]: Might make your logic easier if you used `-Infinity` and `Infinity`

File: src/components/SidePanel/SidePanel.tsx [hunk 3]
```diff
@@ -172,19 +187,23 @@ class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
 
 	handleResize = ({ dX }: { dX: number }): void => {
 		const { startWidth } = this.state;
+		const position = startWidth + dX * (this.props.position === 'right' ? -1 : 1);
+		const adjustedWidth = !_.isNil(this.props.minWidth) && position < this.props.minWidth ? this.props.minWidth
```
Comment (jondlm) [inline src/components/SidePanel/SidePanel.tsx:191]: Take a look at https://lodash.com/docs/4.17.15#clamp and see if it would simplify this.

Comment (pearlanre) [reply src/components/SidePanel/SidePanel.tsx:191]: ah thanks. I thought there was a lodash implementation for this but couldn't find what it was called.

File: src/components/SidePanel/SidePanel.tsx [hunk 4]
```diff
@@ -156,6 +168,8 @@ class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
 		preventBodyScroll: false,
 		topOffset: 0,
 		width: 240,
+		minWidth: -Infinity,
+		maxWidth: window.innerWidth,
```
Comment (jondlm) [inline src/components/SidePanel/SidePanel.tsx:172]: This leaves one final edge case that I can think of. The default `maxWidth` will be pinned to the `maxWidth` when the component mounts (or when the module loads, not sure off the top of my head) which would mean that when the user resizes their browser it wouldn't be taken into account.

Perhaps a better approach would be to set this to `Infinity` and then in the `handleResize` handler check if `this.props.maxWidth === Infinity` and read directly from `window.innerWidth` there. I _think_ that would avoid that edge case.

File: src/components/SidePanel/SidePanel.tsx [hunk 5]
```diff
@@ -172,19 +186,24 @@ class SidePanel extends React.Component<ISidePanelProps, ISidePanelState, {}> {
 
 	handleResize = ({ dX }: { dX: number }): void => {
 		const { startWidth } = this.state;
+		const position = startWidth + dX * (this.props.position === 'right' ? -1 : 1);
+		//const adjustedWidth = !_.isNil(this.props.minWidth) && position < this.props.minWidth ? this.props.minWidth
```
Comment (jondlm) [inline src/components/SidePanel/SidePanel.tsx:190]: Delete this?

**Pull request merged**